### ISSUE TITLE
IngressGateway support for Server First Protocols

### ISSUE THREAD

**Issue by robertlestak** [istio/istio#34591]
We are unable to connect to an SMTP interface running within an Istio-enabled cluster due to the server-first nature of SMTP communication. If we port-forward the service (so, proxying through the Pod's `istio-envoy` container) it all works fine. However when we attempt to connect through the `istio-ingressgateway`, Istio does not establish a connection to the upstream until the client in front of the ingressgateway sends traffic, which is incompatible with SMTP's server-first nature.

We have reviewed the documentation on [Server First Protocols](https://istio.io/latest/docs/ops/deployment/requirements/#server-first-protocols) and all our ports are [explicitly named](https://istio.io/latest/docs/ops/configuration/traffic-management/protocol-selection/#explicit-protocol-selection) `tcp-smtp`.

We also found [this GitHub issue](https://github.com/istio/istio/issues/10832) which appears to be related (Istio not sending traffic to server until client sends first byte), but this issue was closed without a resolution.

[x] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure
[ ] Upgrade

**Expected behavior**

TCP connections to ingressgateway on explicitly labeled `tcp-*` ports should support server-first communication as outlined in the [documentation](https://istio.io/latest/docs/ops/deployment/requirements/#server-first-protocols).

**Steps to reproduce the bug**

# connect through gateway

Connecting to service through ingressgateway fails, expects client to send first bytes, and does not open connection to upstream service until client has sent data. Even when all ports are explicitly labeled as `tcp-smtp`.

## tail ingress gateway logs
```bash
# get name of an ingress gateway pod
POD_NAME=$(kubectl get po -n istio-system -l app=istio-ingressgateway-dev-int -o=name | head -n1)
# enable debug logging
kubectl exec -it -n istio-system $POD_NAME -- curl -X POST http://localhost:15000/logging?level=debug
# tail logs for our SMTP interface
kubectl logs -f -n istio-system $POD_NAME | grep 'cluster outbound|25|v1|nginx-smtp.nginx-dev.svc.cluster.local'
```

## telnet gateway
```bash
# get user-facing hostname of ingressgateway
LB_HOSTNAME=$(kubectl get svc -n istio-system istio-ingressgateway-dev-int -o jsonpath="{.status.loadBalancer.ingress[0].hostname}")
# attempt to connect to the LB
# NOTE: to make this easier for testing, right now we are accepting all hosts on port 25
~ telnet $LB_HOSTNAME 25

Trying [load balancer IP]...
Connected to [load balancer].
Escape character is '^]'.
# NOTE: No server 220 / Ready
# if you look at ingressgateway logs, the ingressgateway has not sent this traffic to the upstream service yet
# hit enter 3 times, then ingressgateway will send traffic to upstream, SMTP will return an 220 followed by a 500 5.5.1 Invalid command, and then you can start your client smtp request
# you will now see in logs that ingressgateway has sent traffic to upstream service

# telnet data
HELO mail.example.com
MAIL FROM:<hello@example.com>
RCPT TO:<hello@example.com>
DATA
From: <hello@example.com>
To: <hello@example.com>
Subject: SMTP test

this is a test
.
QUIT
# end telnet data

# however mail clients expect to be greeted with the server-side 220 before it starts sending data, so this does not work
```

# proxy to service

Port-forwarding the pod's port (traffic still through the pod's respective `istio-proxy` container) works fine, and the server immediately responds with a 220 when the connection is opened.

## port forward nginx service

```bash
k port-forward -n nginx-dev svc/nginx-smtp 2225:25
```

## telnet service
```bash
~ telnet localhost 2225

Trying ::1...
Connected to localhost.
Escape character is '^]'.
220 smtp.example.com ESMTP ready

# telnet data
HELO mail.example.com
MAIL FROM:<hello@example.com>
RCPT TO:<hello@example.com>
DATA
From: <hello@example.com>
To: <hello@example.com>
Subject: SMTP test

this is a test
.
QUIT
# end telnet data
```

**Version** (include the output of `istioctl version --remote` and `kubectl version --short` and `helm version --short` if you used Helm)

```
~ istioctl version --remote
control plane version: 1.6.5-f508fdd78eb0d3444e2bc2b3f36966d904c5db52
data plane version: 1.6.5 (46 proxies)

~ kubectl version --short
Client Version: v1.21.1
Server Version: v1.17.17-eks-087e67
```

**How was Istio installed?**

```
istioctl install -f ./eks_scripts/istio/istio_templates/values-istio-multicluster-gateways.yaml \
                -f ./multi_lb.yaml \
                -f ./eks_scripts/istio/istio_templates/external_prometheus.yaml
```

**Environment where the bug was observed (cloud vendor, OS, etc)**

```
AWS EKS eks.8
```

**Comment by howardjohn** [comment]
What does the Gateway look like? or more importantly the actual envoy config (https://github.com/istio/istio/wiki/Troubleshooting-Istio#collecting-information-2)

**Comment by robertlestak** [comment]
@howardjohn thanks for the prompt response. envoy config dump attached below.

[config_dump.json.txt](https://github.com/istio/istio/files/6956684/config_dump.json.txt)

**Comment by robertlestak** [comment]
@howardjohn thanks for the prompt response. envoy config dump attached below.

[config_dump.json.txt](https://github.com/istio/istio/files/6956717/config_dump.json.txt)

**Comment by howardjohn** [comment]
I could not reproduce like
```

apiVersion: networking.istio.io/v1beta1
kind: VirtualService
metadata:
  name: echo
  namespace: default
spec:
  gateways:
  - echo
  hosts:
  - '*'
  tcp:
  - route:
    - destination:
        host: echo
        port:
          number: 9091

---
apiVersion: networking.istio.io/v1beta1
kind: Gateway
metadata:
  name: echo
  namespace: default
spec:
  selector:
    istio: ingressgateway
  servers:
  - hosts:
    - '*'
    port:
      name: tcp
      number: 1234
      protocol: TCP
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: echo
spec:
  selector:
    matchLabels:
      app: echo
  template:
    metadata:
      labels:
        app: echo
    spec:
      containers:
      - name: echo
        image: gcr.io/istio-testing/app:latest
        imagePullPolicy: IfNotPresent
        args:
        - --tcp=9090
        - --tcp=9091
        - --server-first=9091
        - --port=80
        - --grpc=7070
        - --port=443
        - --tls=443
        - --crt=/cert.crt
        - --key=/cert.key
        env:
        - name: INSTANCE_IP
          valueFrom:
            fieldRef:
              fieldPath: status.podIP
---
apiVersion: v1
kind: Service
metadata:
  name: echo
spec:
  selector:
    app: echo
  ports:
  - name: http
    port: 80
  - name: https
    port: 443
  - name: tcp
    port: 9090
  - name: tcp-sf
    port: 9091
  - name: grpc
    port: 7070
```

Then `client tcp://IP:1234 --server-first`.

Based on this, I think the problem is either in your LoadBalancer provider (EKS), or its something other than "server first" breaking things.

**Comment by robertlestak** [comment]
Thank you for taking the time to test this out.

I did a clean install and reviewed our configs.

It turns out it was Proxy Protocol - we enabled Proxy Protocol on our Gateways to get the clients' real IP (rather than the cloud load balancer IP).

Since we had a proxy-protocol `EnvoyFilter`, that appears to be what was "holding" up the traffic. It was only when I removed the EnvoyFilter but left Proxy Protocol support enabled on the cloud load balancer, I was finally able to get a hint in the SMTP logs:

```
239   SMTP protocol synchronization error (input sent without waiting for greeting): rejected connection from H=[127.0.0.6] input="PROXY TCP4 10.x.x.x.x 10.x.x.x.x 57890 25\r\n"
```

Going to launch a dedicated Gateway with Proxy Protocol support disabled and replace the `type: LoadBalancer` gateway - glad we're able to keep this Istio-native :)

Thank you again and have a great weekend.
